[WMS][12.0] rework stock_location_zone

grindtildeath opened the pull request:
Followup to #653

grindtildeath committed:
Rework stock_location_zone

grindtildeath committed:
Simplify constraint

grindtildeath committed:
Restore copy function

grindtildeath commented on stock_location_zone/models/stock_location.py:
```diff
@@ -3,142 +3,102 @@
-             ".{level:0>2}'")
+    _sql_constraints = [
+        'name_zone_unique',
+        'unique(name, zone_location_id)',
```
Add `company_id`

grindtildeath committed:
fixup! Simplify constraint

grindtildeath commented on stock_location_zone/models/stock_location.py:
```diff
@@ -3,142 +3,112 @@
+                    location.location_id.zone_location_id
+
+    @api.depends('name', 'location_kind', 'location_id.area')
     def _compute_area(self):
```
To move to #713

grindtildeath committed:
fixup! Rework stock_location_zone

grindtildeath committed:
fixup! fixup! Simplify constraint

rousseldenis commented:
How fast reaction :smile: :+1:

Commit:
Remove unique constraint on location's name

It prevents other addons to work properly such as the vertical lift.
We remove the constraint for now, maybe to be re-introduced later in a
more permissive way.

Commit:
Improve performance

* Compute the 'zone' fields in the same method
* Remove the dependency on 'zone_location_id', which seems to take
  a lot of time if not an infinite loop

daramousk commented on stock_location_zone/models/stock_location.py:
```diff
@@ -3,142 +3,95 @@
+             '* Bin: locations without children that are part of a zone'
+             '* Stock: internal locations whose parent is a view'
+             '* Other: any other location',
+    )
```
This could be `store=True` as well, mostly because you cannot group them otherwise.

guewen replied:
Added

daramousk commented on stock_location_zone/views/stock_location.xml:
```diff
@@ -30,17 +21,15 @@
+                <field name="is_zone" />
+                <field name="zone_location_id" />
+                <group expand="0" string="Group By">
+                    <filter string="Location Kind" name="location_kind" domain="[]" context="{'group_by':'location_kind'}"/>
```
This cannot work since the field is not persisted.

daramousk commented on stock_location_zone/models/stock_location.py:
```diff
@@ -3,142 +3,95 @@
+                    break
+                parent = parent.location_id
+            else:
+                zone_location = self.browse()
```
Unessesary if you `zone_location = self` in the beginning.

Commit:
Add area location relation

So we can use it in group-by and filters. Note that we can have several
levels of areas... only the top-level will be used.

This commit fixes the computed field dependencies:
_compute_location_kind should have a dependency on both it's parent's
location_kind and on its child_ids to know if we are in a bin. This
can't work without triggering an infinite loop. The trick used here is
to split the computation of 'zone_location_id + area_location_id' in one
computed method, and move the computation of the kind in a different
method with triggers an the current record's zone_location +
area_location_id, but not on the parent.

Plus the zone_location_id and area_location_id do not depend anymore on
the parent's kind, which is the reason for the infinite loop.

guewen commented:
Reworked the function fields, added `store=True` and added an `area_location_id` similar to `zone_location_id`.

jbaudoux commented on stock_location_zone/readme/DESCRIPTION.rst:
```diff
@@ -1,2 +1,10 @@
+
+Locations are then classified by location kinds that could be:
+
+* Zone: locations that are flagged as being zones
```
```suggestion
* Zone: locations that are flagged as being zones. Zones are subdivisions of the warehouse for splitting picking operations. A picking operator work in a zone and can pick from any location of the zone.
```

jbaudoux commented on stock_location_zone/readme/DESCRIPTION.rst:
```diff
@@ -1,2 +1,10 @@
+Locations are then classified by location kinds that could be:
+
+* Zone: locations that are flagged as being zones
+* Area: locations with children that are part of a zone
```
```suggestion
* Area: locations with children that are part of a zone. Areas are subdivisions of the warehouse for put-away operations. Each area has storage characteristics and strategy, e.g. area for pallets, shelf for boxes...
```

grindtildeath committed:
Apply suggestions from code review

Co-Authored-By: Jacques-Etienne Baudoux <je.baudoux@gmail.com>

jgrandguillaume commented:
/ocabot merge nobump

rousseldenis commented:
/ocabot merge nobump

rousseldenis commented:
/ocabot merge nobump
